File: jetbrains-plugin/src/main/kotlin/com/vscode/jetbrainssync/FileUtils.kt
package com.vscode.jetbrainssync

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.LogicalPosition
import com.intellij.openapi.editor.ScrollType
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileEditor.TextEditor
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.wm.IdeFocusManager
import java.io.File

/**
 * File utility class
 * Provides file operation related utility methods
 */
class FileUtils(private val project: Project, private val log: Logger) {

    /**
     * Check if file is still open in other editors
     */
    fun isFileOpenInOtherTabs(file: VirtualFile): Boolean {
        val fileEditorManager = FileEditorManager.getInstance(project)
        return fileEditorManager.isFileOpen(file)
    }

    /**
     * Determine if it's a regular file editor (only allow regular file systems)
     */
    fun isRegularFile(virtualFile: VirtualFile): Boolean {
        val fileSystem = virtualFile.fileSystem.protocol

        // Whitelist mechanism: only allow regular file system protocols
        val allowedFileSystems = listOf(
            "file"       // Local file system
        )

        return allowedFileSystems.contains(fileSystem)
    }

    /**
     * Get all currently opened file paths
     * Only return regular file editors, filter out special tab windows
     */
    fun getAllOpenedFiles(): List<String> {
        val fileEditorManager = FileEditorManager.getInstance(project)
        return fileEditorManager.openFiles
            .filter { virtualFile ->
                // Only keep regular file editors, filter out all special tab windows
                isRegularFile(virtualFile)
            }
            .map { it.path }
    }


    /**
     * Extract filename from file path
     * @param filePath File path
     * @return Filename
     */
    fun extractFileName(filePath: String): String {
        return File(filePath).name
    }

    /**
     * Get file path of editor
     * @param virtualFile Virtual file
     * @return File path
     */
    fun getVirtualFilePath(virtualFile: VirtualFile): String {
        return virtualFile.path
    }

    /**
     * Get cursor position of editor
     * @param editor Text editor
     * @return Cursor position Pair<line number, column number>
     */
    fun getEditorCursorPosition(editor: Editor): Pair<Int, Int> {
        val position = editor.caretModel.logicalPosition
        return Pair(position.line, position.column)
    }

    /**
     * Get selection range coordinates of editor
     * @param editor Text editor
     * @return Selection range coordinates (startLine, startColumn, endLine, endColumn), return null if no selection
     */
    fun getSelectionCoordinates(editor: Editor): Quadruple<Int, Int, Int, Int>? {
        val selectionModel = editor.selectionModel
        val hasSelection = selectionModel.hasSelection()

        return if (hasSelection) {
            val startPosition = editor.offsetToLogicalPosition(selectionModel.selectionStart)
            val endPosition = editor.offsetToLogicalPosition(selectionModel.selectionEnd)
            Quadruple(startPosition.line, startPosition.column, endPosition.line, endPosition.column)
        } else {
            null
        }
    }

    /**
     * Quadruple data class, used to return four coordinates of selection range
     */
    data class Quadruple<out A, out B, out C, out D>(
        val first: A,
        val second: B,
        val third: C,
        val fourth: D
    )


    /**
     * Get currently selected file and editor
     * @return Pair<TextEditor?, VirtualFile?> combination of editor and virtual file
     */
    fun getCurrentActiveEditorAndFile(): Pair<Editor?, VirtualFile?> {
        val fileEditorManager = FileEditorManager.getInstance(project)
        val selectedEditor = fileEditorManager.selectedTextEditor
        val selectedFile = fileEditorManager.selectedFiles.firstOrNull()
        if (selectedFile != null && isRegularFile(selectedFile)) {
            return Pair(selectedEditor, selectedFile)
        }
        return Pair(null, null);
    }

    /**
     * Determine if current editor has focus
     * @return Boolean whether current editor has focus
     */
    fun isEditorFocused(): Boolean {
        val focusManager = IdeFocusManager.getInstance(project)
        val focusOwner = focusManager.focusOwner

        // Get current active editor
        val (currentEditor, _) = getCurrentActiveEditorAndFile()

        // If there is no active editor, it definitely doesn't have focus
        if (currentEditor == null) {
            return false
        }

        // Check if current focus component belongs to editor
        return focusOwner != null && currentEditor.contentComponent.isAncestorOf(focusOwner)
    }

    /**
     * Close file by file path
     * If direct path matching fails, will try matching by filename
     */
    fun closeFileByPath(filePath: String) {
        try {
            log.info("Preparing to close file: $filePath")
            val fileEditorManager = FileEditorManager.getInstance(project)
            // Try matching by file path
            val virtualFile = findFileByPath(filePath)
            virtualFile?.let { vFile ->
                if (fileEditorManager.isFileOpen(vFile)) {
                    fileEditorManager.closeFile(vFile)
                    log.info("✅ Successfully closed file: $filePath")
                    return
                } else {
                    log.warn("⚠️ File not open, no need to close: $filePath")
                    return
                }
            }
            log.warn("❌ File not found: $filePath")
        } catch (e: Exception) {
            log.warn("Failed to close file: $filePath - ${e.message}", e)
        }
    }

    /**
     * Open file by file path
     * Supports opening newly created files from other IDEs, solving file not found issues by refreshing VFS cache
     * @param filePath File path
     * @param focusEditor Whether to get focus, default is true
     * @return Returns opened TextEditor, returns null if failed
     */
    fun openFileByPath(filePath: String, focusEditor: Boolean = true): TextEditor? {
        try {
            log.info("Preparing to open file: $filePath")
            val fileEditorManager = FileEditorManager.getInstance(project)
            // Try finding virtual file by file path
            val virtualFile = findFileByPath(filePath)
            virtualFile?.let { vFile ->
                // FileEditorManager.openFile() will automatically reuse opened files, no need for manual check
                val editors = fileEditorManager.openFile(vFile, focusEditor)
                val editor = editors.firstOrNull() as? TextEditor

                if (editor != null) {
                    log.info("✅ Successfully opened file: $filePath")
                    return editor
                } else {
                    log.warn("❌ Unable to get file editor: $filePath")
                    return null
                }
            }
            log.warn("❌ File not found: $filePath")
            return null
        } catch (e: Exception) {
            log.warn("Failed to open file: $filePath - ${e.message}", e)
            return null
        }
    }

    /**
     * Find virtual file, refresh VFS cache and retry if not found
     * This method specifically handles synchronization issues of newly created files in other IDEs
     * @param filePath File path
     * @return Virtual file object, returns null if not found
     */
    fun findFileByPath(filePath: String): VirtualFile? {
        val file = File(filePath)
        val fileSystem = LocalFileSystem.getInstance()
        // First attempt: direct search
        var virtualFile = fileSystem.findFileByIoFile(file)
        if (virtualFile != null) {
            log.info("Found file directly: ${file.path}")
            return virtualFile
        }

        log.info("File not found, starting VFS cache refresh: ${file.path}")

        // Second attempt: search after refreshing parent directory
        val parentFile = file.parentFile
        if (parentFile != null && parentFile.exists()) {
            val parentVirtualFile = fileSystem.findFileByIoFile(parentFile)
            parentVirtualFile?.refresh(false, true)
            log.info("Refreshed parent directory VFS cache: ${parentFile.path}")

            virtualFile = fileSystem.findFileByIoFile(file)
            if (virtualFile != null) {
                log.info("Found file after refreshing parent directory: ${file.path}")
                return virtualFile
            }
        }

        // Third attempt: search after forcing refresh of entire file system
        log.info("Refreshing parent directory ineffective, performing global VFS refresh")
        fileSystem.refresh(false)

        // Give file system some time to update index
        try {
            Thread.sleep(100)
        } catch (e: InterruptedException) {
            Thread.currentThread().interrupt()
        }

        virtualFile = fileSystem.findFileByIoFile(file)
        if (virtualFile != null) {
            log.info("Found file after global refresh: ${file.path}")
            return virtualFile
        }

        log.warn("All refresh attempts failed, file may not exist: ${file.path}")
        return null
    }


    /**
     * Unified handling of selection and cursor movement
     * First handle selection state (set selection if there is one, clear selection if none), then ensure cursor position is within visible area
     * @param textEditor Text editor
     * @param line Cursor line number
     * @param column Cursor column number
     * @param startLine Selection start line number (optional)
     * @param startColumn Selection start column number (optional)
     * @param endLine Selection end line number (optional)
     * @param endColumn Selection end column number (optional)
     */
    fun handleSelectionAndNavigate(
        textEditor: TextEditor,
        line: Int,
        column: Int,
        startLine: Int? = null,
        startColumn: Int? = null,
        endLine: Int? = null,
        endColumn: Int? = null
    ) {
        try {
            log.info("Preparing to handle selection and cursor navigation: ${LogFormatter.cursorLog(line, column)}, ${LogFormatter.selectionLog(startLine, startColumn, endLine, endColumn)}")

            ApplicationManager.getApplication().runWriteAction {
                val selectionModel = textEditor.editor.selectionModel

                // First handle selection state
                if (startLine != null && startColumn != null && endLine != null && endColumn != null) {
                    // Has selection range, set selection
                    val startPosition = LogicalPosition(startLine, startColumn)
                    val endPosition = LogicalPosition(endLine, endColumn)

                    selectionModel.setSelection(
                        textEditor.editor.logicalPositionToOffset(startPosition),
                        textEditor.editor.logicalPositionToOffset(endPosition)
                    )
                    log.info("✅ Successfully set selection range: ${LogFormatter.selection(startLine, startColumn, endLine, endColumn)}")
                } else {
                    // No selection range, clear selection
                    selectionModel.removeSelection()
                    log.info("✅ Successfully cleared selection state, ${LogFormatter.cursorLog(line, column)}")
                }

                // Then move cursor to specified position
                val cursorPosition = LogicalPosition(line, column)
                textEditor.editor.caretModel.moveToLogicalPosition(cursorPosition)
                log.info("✅ Successfully moved cursor to position: ${LogFormatter.cursor(line, column)}")

                // Ensure cursor position is within visible area
                val visibleArea = textEditor.editor.scrollingModel.visibleArea
                val targetPoint = textEditor.editor.logicalPositionToXY(cursorPosition)

                if (!visibleArea.contains(targetPoint)) {
                    textEditor.editor.scrollingModel.scrollToCaret(ScrollType.MAKE_VISIBLE)
                    log.info("✅ Cursor position not visible, scrolled to: ${LogFormatter.cursor(line, column)}")
                } else {
                    log.info("Cursor position already within visible area, no scrolling needed")
                }
            }
            log.info("✅ Selection and cursor navigation processing completed")
        } catch (e: Exception) {
            log.warn("❌ Failed to handle selection and cursor navigation: ${LogFormatter.cursorLog(line, column)} - ${e.message}", e)
        }
    }
}

File: vscode-extension/src/FileUtils.ts
import * as vscode from 'vscode';
import * as path from 'path';
import {Logger} from './Logger';
import {LogFormatter} from './Type';

/**
 * File utility class
 * Provides file operation related utility methods
 */
export class FileUtils {
    private static logger: Logger;

    /**
     * Initialize utility class
     * @param logger Logger instance
     */
    static initialize(logger: Logger): void {
        this.logger = logger;
    }

    /**
     * Check if file is still open in other tabs
     */
    static isFileOpenInOtherTabs(filePath: string): boolean {
        for (const tabGroup of vscode.window.tabGroups.all) {
            for (const tab of tabGroup.tabs) {
                if (FileUtils.isRegularFileTab(tab)) {
                    const uri = (tab.input as vscode.TabInputText).uri;
                    if (uri.fsPath === filePath) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    /**
     * Check if it's a regular file tab (only allows regular file protocols)
     */
    static isRegularFileTab(tab: vscode.Tab): boolean {
        const input = tab.input;

        // Only accept TabInputText type, exclude all other types
        if (!(input instanceof vscode.TabInputText)) {
            return false;
        }

        const uri = input.uri;

        // Reuse isRegularFileUri logic
        return this.isRegularFileUri(uri);
    }

    /**
     * Check if it's a regular file URI (only allows regular file protocols)
     */
    static isRegularFileUri(uri: vscode.Uri): boolean {
        // Whitelist mechanism: only allow regular file protocols
        const allowedSchemes = [
            'file'              // Local file system
        ];

        return allowedSchemes.includes(uri.scheme);
    }

    /**
     * Check if editor is a valid regular file editor
     * @param editor Text editor
     * @returns Whether it's a regular file editor
     */
    static isRegularFileEditor(editor: vscode.TextEditor): boolean {
        return this.isRegularFileUri(editor.document.uri);
    }

    /**
     * Get all currently opened file paths
     * Only return regular file tabs, filter out special tab windows
     */
    static getAllOpenedFiles(): string[] {
        const openedFiles: string[] = [];

        for (const tabGroup of vscode.window.tabGroups.all) {
            for (const tab of tabGroup.tabs) {
                // Only handle regular text file tabs, filter out all special tab types
                if (this.isRegularFileTab(tab)) {
                    const tabInput = tab.input as vscode.TabInputText;
                    const uri = tabInput.uri;

                    // File protocol already validated in isRegularFileTab, add directly
                    openedFiles.push(uri.fsPath);
                }
            }
        }

        return openedFiles;
    }


    /**
     * Extract file name from file path
     * @param filePath File path
     * @returns File name
     */
    static extractFileName(filePath: string): string {
        return path.basename(filePath);
    }

    /**
     * Get file path of editor
     * @param editor Text editor
     * @returns File path
     */
    static getEditorFilePath(editor: vscode.TextEditor): string {
        return editor.document.uri.fsPath;
    }

    /**
     * Get cursor position of editor
     * @param editor Text editor
     * @returns Cursor position {line: number, column: number}
     */
    static getEditorCursorPosition(editor: vscode.TextEditor): { line: number, column: number } {
        const position = editor.selection.active;
        return {
            line: position.line,
            column: position.character
        };
    }

    /**
     * Get selection range coordinates of editor
     * @param editor Text editor
     * @returns Selection range coordinates {startLine, startColumn, endLine, endColumn}, returns null if no selection
     */
    static getSelectionCoordinates(editor: vscode.TextEditor): { startLine: number, startColumn: number, endLine: number, endColumn: number } | null {
        const selection = editor.selection;
        const hasSelection = !selection.isEmpty;

        if (hasSelection) {
            return {
                startLine: selection.start.line,
                startColumn: selection.start.character,
                endLine: selection.end.line,
                endColumn: selection.end.character
            };
        } else {
            return null;
        }
    }

    /**
     * Get current active editor
     * Use multi-layer strategy to ensure getting the current editing area's text editor even when focus is not on the editor
     * @returns Returns the currently active TextEditor, returns null if none
     */
    static getCurrentActiveEditor(): vscode.TextEditor | null {
        // Strategy 1: Prioritize using activeTextEditor (when focus is on editor)
        const activeEditor = vscode.window.activeTextEditor;
        if (activeEditor && this.isRegularFileEditor(activeEditor)) {
            this.logger.info(`Got editor via activeTextEditor: ${this.extractFileName(activeEditor.document.uri.fsPath)}`);
            return activeEditor;
        }

        // Strategy 2: Get from visible editors (fallback when focus is not on editor)
        const visibleEditors = vscode.window.visibleTextEditors;
        if (visibleEditors.length > 0) {
            // Find regular file editors, prioritize selecting the last one (usually the most recently active)
            const regularEditors = visibleEditors.filter(editor => this.isRegularFileEditor(editor));
            if (regularEditors.length > 0) {
                const selectedEditor = regularEditors[regularEditors.length - 1];
                this.logger.info(`Got editor via visible editors: ${this.extractFileName(selectedEditor.document.uri.fsPath)}`);
                return selectedEditor;
            }
        }

        // Strategy 3: Get editor corresponding to active tab from active tab group
        const activeTabGroup = vscode.window.tabGroups.activeTabGroup;
        if (activeTabGroup && activeTabGroup.activeTab) {
            const activeTab = activeTabGroup.activeTab;
            if (this.isRegularFileTab(activeTab)) {
                const tabInput = activeTab.input as vscode.TabInputText;
                const uri = tabInput.uri;

                // Find corresponding editor
                const correspondingEditor = vscode.window.visibleTextEditors.find(
                    editor => editor.document.uri.toString() === uri.toString()
                );
                if (correspondingEditor && this.isRegularFileEditor(correspondingEditor)) {
                    this.logger.info(`Got editor via active tab group: ${this.extractFileName(uri.fsPath)}`);
                    return correspondingEditor;
                }
            }
        }

        // Strategy 4: Find the most recent regular file tab from all tab groups
        for (const tabGroup of vscode.window.tabGroups.all) {
            if (tabGroup.activeTab && this.isRegularFileTab(tabGroup.activeTab)) {
                const tabInput = tabGroup.activeTab.input as vscode.TabInputText;
                const uri = tabInput.uri;

                // Find corresponding editor
                const correspondingEditor = vscode.window.visibleTextEditors.find(
                    editor => editor.document.uri.toString() === uri.toString()
                );
                if (correspondingEditor && this.isRegularFileEditor(correspondingEditor)) {
                    this.logger.info(`Got editor via tab group: ${this.extractFileName(uri.fsPath)}`);
                    return correspondingEditor;
                }
            }
        }

        this.logger.warn('Could not get any active editor, possibly no text files are open');
        return null;
    }

    /**
     * Check if current editor has focus
     * @returns Whether current editor has focus
     */
    static isEditorFocused(): boolean {
        // In VSCode, activeTextEditor only has a value when the editor has focus
        // If activeTextEditor exists and is a regular file editor, it means the editor has focus
        const activeEditor = vscode.window.activeTextEditor;
        if (activeEditor && this.isRegularFileEditor(activeEditor)) {
            this.logger.info(`Editor has focus: ${this.extractFileName(activeEditor.document.uri.fsPath)}`);
            return true;
        }

        // If activeTextEditor doesn't exist or is not a regular file editor, it means focus is elsewhere
        this.logger.info('Editor does not have focus, focus might be on other tool windows or panels');
        return false;
    }


    /**
     * Close file by file path
     * Uses two-phase close strategy: try tab method first, then textDocument method if it fails
     * Only uses exact path matching, not filename matching
     */
    static async closeFileByPath(filePath: string): Promise<void> {
        try {
            this.logger.info(`Preparing to close file: ${filePath}`);

            // Phase 1: Try to close via tabGroups API
            let targetTab: vscode.Tab | undefined;

            for (const tabGroup of vscode.window.tabGroups.all) {
                for (const tab of tabGroup.tabs) {
                    if (this.isRegularFileTab(tab)) {
                        const tabInput = tab.input as vscode.TabInputText;
                        if (tabInput.uri.fsPath === filePath) {
                            targetTab = tab;
                            break;
                        }
                    }
                }
                if (targetTab) break;
            }

            if (targetTab) {
                try {
                    await vscode.window.tabGroups.close(targetTab);
                    this.logger.info(`✅ Successfully closed file via tab method: ${filePath}`);
                    return;
                } catch (tabCloseError) {
                    this.logger.warn(`Tab method close failed, trying fallback: ${filePath}`, tabCloseError as Error);
                }
            } else {
                this.logger.warn(`❌ File not found in tabs: ${filePath}`);
            }

            // Phase 2: Fallback - use original textDocument method to close
            this.logger.info(`🔄 Trying to close via textDocument method: ${filePath}`);
            const editorToClose = vscode.workspace.textDocuments.find(doc => doc.uri.fsPath === filePath);

            if (editorToClose) {
                await vscode.window.showTextDocument(editorToClose);
                await vscode.commands.executeCommand('workbench.action.closeActiveEditor');
                this.logger.info(`✅ Successfully closed file via textDocument method: ${filePath}`);
            } else {
                this.logger.warn(`❌ File also not found in textDocument: ${filePath}`);
            }
        } catch (error) {
            this.logger.warn(`Failed to close file: ${filePath}`, error as Error);
        }
    }

    /**
     * Open file by file path
     * @param filePath File path
     * @param focusEditor Whether to get focus, defaults to true
     * @returns Returns the opened TextEditor, returns null if failed
     */
    static async openFileByPath(filePath: string, focusEditor: boolean = true): Promise<vscode.TextEditor | null> {
        try {
            this.logger.info(`Preparing to open file: ${filePath}`);
            const uri = vscode.Uri.file(filePath);
            const document = await vscode.workspace.openTextDocument(uri);
            const editor = await vscode.window.showTextDocument(document, {preview: false, preserveFocus: !focusEditor});
            this.logger.info(`✅ Successfully opened file: ${filePath}`);
            return editor;
        } catch (error) {
            this.logger.warn(`Failed to open file: ${filePath}`, error as Error);
            return null;
        }
    }


    /**
     * Unified handling of selection and cursor movement
     * Supports cursor at any position, not limited by selection range
     * @param editor Text editor
     * @param line Cursor line number
     * @param column Cursor column number
     * @param startLine Selection start line number ( optional)
     * @param startColumn Selection start column number ( optional)
     * @param endLine Selection end line number ( optional)
     * @param endColumn Selection end column number ( optional)
     */
    static handleSelectionAndNavigate(
        editor: vscode.TextEditor,
        line: number,
        column: number,
        startLine?: number,
        startColumn?: number,
        endLine?: number,
        endColumn?: number
    ): void {
        try {
            this.logger.info(`Preparing to handle selection and cursor navigation: ${LogFormatter.cursorLog(line, column)}, ${LogFormatter.selectionLog(startLine, startColumn, endLine, endColumn)}`);

            const cursorPosition = new vscode.Position(line, column);

            // Check if there are valid selection range parameters
            const hasValidSelection = startLine !== undefined && startColumn !== undefined &&
                endLine !== undefined && endColumn !== undefined;

            // Check if it's a valid non-zero length selection
            const hasNonZeroSelection = hasValidSelection &&
                !(startLine === endLine && startColumn === endColumn);

            if (hasNonZeroSelection) {
                // Handle valid selection range, need to correctly set cursor position
                const startPosition = new vscode.Position(startLine, startColumn);
                const endPosition = new vscode.Position(endLine, endColumn);

                // Determine selection direction by distance: calculate distance from cursor to selection start and end
                // If cursor is closer to start, it means selection is from bottom to top (anchor at end)
                // If cursor is closer to end, it means selection is from top to bottom (anchor at start)
                const distanceToStart = Math.abs((line - startLine) * 1000 + (column - startColumn));
                const distanceToEnd = Math.abs((line - endLine) * 1000 + (column - endColumn));

                if (distanceToStart < distanceToEnd) {
                    // Cursor is closer to start position, selecting from bottom to top
                    // VSCode Selection constructor: new Selection(anchor, active)
                    // anchor is the selection anchor point, active is the actual cursor position
                    editor.selection = new vscode.Selection(endPosition, cursorPosition);
                    this.logger.info(`✅ Successfully set selection range (bottom to top): ${LogFormatter.selection(startLine, startColumn, endLine, endColumn)}, ${LogFormatter.cursorLog(line, column)}`);
                } else {
                    // Cursor is closer to end position, selecting from top to bottom
                    editor.selection = new vscode.Selection(startPosition, cursorPosition);
                    this.logger.info(`✅ Successfully set selection range (top to bottom): ${LogFormatter.selection(startLine, startColumn, endLine, endColumn)}, ${LogFormatter.cursorLog(line, column)}`);
                }
            } else {
                // Clear selection state, only set cursor position
                editor.selection = new vscode.Selection(cursorPosition, cursorPosition);
                this.logger.info(`✅ Successfully cleared selection state, ${LogFormatter.cursorLog(line, column)}`);
            }

            // Ensure cursor position is within visible area
            const visibleRange = editor.visibleRanges[0];
            if (!visibleRange || !visibleRange.contains(cursorPosition)) {
                editor.revealRange(
                    new vscode.Range(cursorPosition, cursorPosition),
                    vscode.TextEditorRevealType.InCenter
                );
                this.logger.info(`✅ Cursor position not visible, scrolled to: ${LogFormatter.cursor(line, column)}`);
            } else {
                this.logger.info(`Cursor position already within visible area, no scrolling needed`);
            }

            this.logger.info(`✅ Selection and cursor navigation processing completed`);
        } catch (error) {
            this.logger.warn(`❌ Failed to handle selection and cursor navigation: ${LogFormatter.cursorLog(line, column)}`, error as Error);
        }
    }

} 